Update CLI docs

TejasQ opened the pull request:
This PR updates the docs for the CLI a bit. These will also be consumed by docs.xata.io to reduce the maintenance burden and prevent the global docs site from going stale.

Preview: https://github.com/xataio/client-ts/blob/cli-docs/cli/README.md
Docs site preview: https://docs-git-cli-xata.vercel.app/cli/getting-started

There are some links that are expected to 404 until https://github.com/xataio/docs/pull/185 is merged.

gimenete commented on cli/README.md:
````diff
@@ -1,55 +1,74 @@
+To install the Xata CLI globally, run the following command:
+
+```sh
+npm i --location global @xata.io/cli
````
```suggestion
npm install -g @xata.io/cli
```

I think the `--location global` part is not very common. I had never seen it before.

TejasQ replied:
It's the current preferred way to install globally:

<img width="1011" alt="image" src="https://user-images.githubusercontent.com/9947422/177975111-3a148555-690d-472d-a4b2-f5039f26ad71.png">

gimenete commented on cli/README.md:
````diff
@@ -1,55 +1,74 @@
 
-```bash
-npm install @xata.io/cli -g
+Our CLI is distributed both as an **[npm](https://www.npmjs.com/package/@xata.io/cli) package** for developers who work with Node.js, and a **cross-platform binary** for those who don't. Below, you'll find commands that you can copy and paste into your platform's shell that will help you install the CLI on your specific platform.
````
> developers who work with Node.js

Not sure about that phrasing, because one might be a web developer, not using Node.js at all directly, and have npm installed.

TejasQ replied:
Thanks for this. Do you have a suggestion? How about this:

```suggestion
Our CLI is distributed both as an **[npm](https://www.npmjs.com/package/@xata.io/cli) package** for JavaScript developers, and a **cross-platform binary** for those who don't. Below, you'll find commands that you can copy and paste into your platform's shell that will help you install the CLI on your specific platform.
```

TejasQ replied:
Or even this:


```suggestion
Our CLI is distributed both as an **[npm](https://www.npmjs.com/package/@xata.io/cli) package** for developers who have Node.js installed on their systems, and a **cross-platform binary** for those who don't. Below, you'll find commands that you can copy and paste into your platform's shell that will help you install the CLI on your specific platform.
```

TejasQ replied:
Hmmm, based on https://github.com/xataio/client-ts/pull/400#pullrequestreview-1032729227, this maybe a bit short-sighted. How about:

```suggestion
Our CLI is distributed as an **[npm](https://www.npmjs.com/package/@xata.io/cli) package**. Here's how you can install it.
```

TejasQ replied:
Done in https://github.com/xataio/client-ts/pull/400/commits/137212cf19d8d42871a64ffac6acaa398d52ef7c

gimenete commented on cli/README.md:
````diff
@@ -1,55 +1,74 @@
 ```
 
-Now you can just run `xata --help` to see all the available commands.
+This will install the Xata CLI. We recommend installing it globally because it becomes much more convenient to work with, having you invoke a `xata` command instead of the more convoluted alternative with a project-scoped installation (`npm run xata`), or an npx-based installation (`npx @xata.io/cli`). The alternative approaches also work depending on your preference.
````
Nitpick: `alternative` is repeated in two sentences very close.

gimenete commented on cli/README.md:
```diff
@@ -1,55 +1,74 @@
+
+#### Windows
+
+Paste the following command into your terminal and hit enter.
```
```suggestion
Paste the following command into your terminal and hit Enter.
```

For consisstency with the previous section.

TejasQ replied:
Thanks a lot!

TejasQ replied:
Done in https://github.com/xataio/client-ts/pull/400/commits/ffc85ef36adff7dab19618f707613e1d01adb70a

gimenete commented on cli/README.md:
```diff
@@ -1,55 +1,74 @@
+
+Paste the following command into your terminal and hit enter.
 
-Almost all commands can work in two different modes: globally or scoped to a project. You can configure a project with `xata init`. This will create a `.xatarc` file (although `.xatarc.json` and `package.json` are also supported) and an `.env` file.
```
I think explaining the difference between the two modes (global vs project) is important.

TejasQ replied:
Done in https://github.com/xataio/client-ts/pull/400/commits/9778eeba7b810f0c4ca53f9fdafee2af990bce9d

gimenete commented on cli/README.md:
````diff
@@ -1,55 +1,74 @@
+iwr https://xata.io/install.ps1 -useb | iex
+```
 
-- The project configuration file contains some preferences and information, including the `databaseURL`.
-- The `.env` file contains an API key and optionally a fallback branch name (more about the fallback branch in the [Git integration](#git-integration) section).
````
Same here, I think this information is important.

TejasQ replied:
Done in https://github.com/xataio/client-ts/pull/400/commits/9778eeba7b810f0c4ca53f9fdafee2af990bce9d

gimenete commented on cli/README.md:
```diff
@@ -1,55 +1,74 @@
-- The `.env` file contains an API key and optionally a fallback branch name (more about the fallback branch in the [Git integration](#git-integration) section).
+## Authentication
 
-When working in _global mode_, commands won't know which database you want to use, because there's no project configuration file, and thus, no `databaseURL` configured. So, you can either pass a `--db` flag to specify the database URL, or just don't pass any argument and you will be prompted interactively to select a workspace, then a database, and depending on the command, a branch name.
```
And same here. Many commands work differently depending if you are in a project or not.

TejasQ replied:
Done in https://github.com/xataio/client-ts/pull/400/commits/9778eeba7b810f0c4ca53f9fdafee2af990bce9d

gimenete commented on cli/README.md:
```diff
@@ -1,55 +1,74 @@
+## Authentication
 
-When working in _global mode_, commands won't know which database you want to use, because there's no project configuration file, and thus, no `databaseURL` configured. So, you can either pass a `--db` flag to specify the database URL, or just don't pass any argument and you will be prompted interactively to select a workspace, then a database, and depending on the command, a branch name.
+To authenticate your CLI installation, run `xata auth login`. Here, you will be presented with a couple of options:
```
`xata auth login` is not required if you have a project configured, because the CLI will use the API key you have in the `.env` file in your project.

So, `xata auth login` is optional and it's important to clarify that.

TejasQ replied:
Thanks a lot for this. I was writing this more through the lens of “getting started”, but this is an excellent point about existing projects.

TejasQ replied:
Done in https://github.com/xataio/client-ts/pull/400/commits/ecd76a20e05ba5c96448701b3a4e5eae461da85c

gimenete commented on cli/README.md:
```diff
@@ -1,55 +1,74 @@
+- if you don't have a [workspace](https://docs.xata.io/concepts/workspaces), you will be prompted to create one.
+- if you have workspaces, you'll be asked to choose one, or to create a new one.
 
-- It will create a project configuration file (`.xatarc`).
-- It will create or update your `.env` file to store a database-scoped API key. (At the time of writing these docs it'll actually store a personal API key, but this will change soon).
-- It will ask you if you want to install the SDK and/or use the TypeScript/JavaScript code generator.
```
I think this information was very relevant.

TejasQ replied:
Done in https://github.com/xataio/client-ts/pull/400/commits/b463f7b00fb60d1da95d400bb457ce1749ba5534

gimenete commented on cli/README.md:
```diff
@@ -1,55 +1,74 @@
+## Code Generation
 
-Code generation will be executed automatically in some cases by the CLI (such as when creating a new Xata branch). You can always invoke it directly with `xata codegen`.
+We strongly recommend using the CLI to generate your Xata fetcher (sometimes called a _client_) for you. This will help make your queries to Xata predictable and safe, while also providing autocomplete and safety.
```
We are using the naming `XataClient` in other places. I prefer that than `Xata fetcher`. Specially because `fetcher` sounds like you can only fetch data, and not mutate it. Besides, we are offering a fully fledged ORM, `fetcher` feels like something small, like a utility function.

TejasQ replied:
Thanks, I hear that. The intent is more to speak to developers who are familiar with `fetch` (a lot) speaking their language and making a mental association with how it works: `fetch` can be used to both read and mutate data, just like the client. 

Meanwhile, my fear of using just the term “client” with no mental association otherwise is that we use “client” in multiple ways:

- a client application to a remote server, like the Xata Web UI
- the Spotify app on a smart TV is a Spotify client
- what we call a client here 

Using the term `fetch`, then clarifying straight after that this is sometimes called a client, is our way of clarifying our intent with the client word. 

Does that make sense? Or should we still change it like this do you think?

gimenete commented on cli/README.md:
```diff
@@ -1,55 +1,74 @@
+We strongly recommend using the CLI to generate your Xata fetcher (sometimes called a _client_) for you. This will help make your queries to Xata predictable and safe, while also providing autocomplete and safety.
 
-If you make changes to the schema through the UI or the REST API, you will need to run `xata codegen` to update your local generated code with the new changes.
+Once you have a generated fetcher, you can use import it into your project and using it. The `.` character (object property access in JavaScript) will provide autocompletion hints. You can now also use the [SDK Playground](https://docs.xata.io/sdk/playground) in the Xata web interface to test out your queries and then copy and paste them into your applications.
```
> you can use import it into your project and using it. The `.` character (object property access in JavaScript) will provide autocompletion hints

This part is confusing.

TejasQ replied:
Thanks a lot! Will think of a solution. 🧠 ⏳

TejasQ replied:
Done in https://github.com/xataio/client-ts/pull/400/commits/1e8f128da51b08b1a9b8598c4fe2baf2f857a4f6

tsg commented on cli/README.md:
```diff
@@ -1,55 +1,74 @@
+
+### Installing the Binary without npm
+
+Our CLI can also be installed as a cross-platform binary for our users who do not have `npm` installed on their systems. Depending on your platform of choice, here are installation instructions to get you started.
```
To be sure, this is not yet available. The below instructions are currently installing the legacy Go CLI.

I was thinking to remove or hide this section until we add back the binary install option, otherwise we're going to get trouble with people installing the old CLI.

TejasQ replied:
Sorry, I missed this: https://xata-hq.slack.com/archives/C032EKCBUGM/p1656916087379719?thread_ts=1656873846.216119&cid=C032EKCBUGM

Let's remove it. 💪

TejasQ replied:
Done in https://github.com/xataio/client-ts/pull/400/commits/ffc85ef36adff7dab19618f707613e1d01adb70a

TejasQ commented on cli/README.md:
```diff
@@ -62,17 +81,21 @@ graph TD
+- otherwise, if a Xata branch with the same name as the git branch exists, we use that.
 
-If none of them apply, a fallback branch is used. This is useful to avoid connecting to your production database. If you are developing a feature that doesn't require schema changes, you'll have a git branch, but not an associated xata branch. In this case to prevent using the default xata branch, that could be production, you can specify in your `.env` file a `XATA_FALLBACK_BRANCH`. You can have ...
+If none of them apply, the **fallback branch** is used. This is a safety measure to avoid connecting to your production database. If you are developing a feature that doesn't require schema changes, you'll have a git branch, but not an associated Xata branch. In this case to prevent using the default Xata branch–which could be production– we allow specifying a `XATA_FALLBACK_BRANCH` in your local ...
```
Forgive me—what exactly is the suggestion here? Is it about spacing around the em-dash?

As far as I know in English, there are no hard rules about this. Sometimes publications either pad em-dashes with spaces, but others don’t depending on literary preferences. 

So far, we’ve opted to omit them for the same reasons Merriam-Webster did (ref: https://www.merriam-webster.com/words-at-play/em-dash-en-dash-how-to-use) but since there’s no hard rules, I don’t have a strong opinion. Happy to have a discussion and a proposal around this if you like, but changing it here in this case will make the rest of the docs and other written media inconsistent and probably is a bigger task.

gimenete replied:
Cool. I didn't know it can be used without spaces. I've always seen it with spaces, I think. To me it's easier to read, but not strong opinions.

Nitpick: the character to use seems to be — not - acording to that article.

TejasQ replied:
Fuck, you're right. These are en-dashes, not em-dashes. Great catch! Thanks! Fixed in 75e5b6c6ecd2459e69368c3b7461d6ee64bc155d

TejasQ commented on cli/README.md:
```diff
@@ -1,55 +1,74 @@
-When working in _global mode_, commands won't know which database you want to use, because there's no project configuration file, and thus, no `databaseURL` configured. So, you can either pass a `--db` flag to specify the database URL, or just don't pass any argument and you will be prompted interactively to select a workspace, then a database, and depending on the command, a branch name.
+To authenticate your CLI installation, run `xata auth login`. Here, you will be presented with a couple of options:
 
-In order to configure authentication for the _global mode_, you have to authenticate first using `xata auth login`. This will guide you through the process to generate a new global API key that will be stored in `~/.config/xata/credentials`.
```
Sorry if I missed this in the original README–how does one configure authentication for the project mode?

gimenete replied:
In project mode an `.env` file is created with an API key. The CLI uses that env file if it exists and there's no need for a global authentication. The idea is to make that API key scoped to the database, so it's also safer than a global API key.

Only for workspace operations you'd need to use a global API key. Once we have database-scoped API keys, the CLI will error with specific messages if you don't have a global API key when doing workspace operations.

TejasQ commented on cli/README.md:
```diff
@@ -1,55 +1,74 @@
+## Initializing a Project
 
-In ordert to configure the CLI in a project you'll run `xata init --db=https://{workspace}.xata.sh/db/{database}`. The flag `--db` is not mandatory but recommended, because this way you won't be asked for the workspace/database and thus, a global API key is not required.
+To initialize a new database, run `xata init` in your project's directory.
```
Done in https://github.com/xataio/client-ts/pull/400/commits/d709edbb6812c3a4e82508113fe0e8d4ab82de93

TejasQ commented on cli/README.md:
```diff
@@ -1,55 +1,74 @@
-## Configuring a project
+## Initializing a Project
 
-In ordert to configure the CLI in a project you'll run `xata init --db=https://{workspace}.xata.sh/db/{database}`. The flag `--db` is not mandatory but recommended, because this way you won't be asked for the workspace/database and thus, a global API key is not required.
```
I think I clarified this in https://github.com/xataio/client-ts/pull/400/commits/9778eeba7b810f0c4ca53f9fdafee2af990bce9d, but let me know if it's not enough and I can do more.

TejasQ commented:
Thanks for the awesome and thorough reviews everyone! I've addressed all the comments, though I haven't yet read it all as one piece over myself. It's 6pm, so I will do this as the first thing on Monday. Thanks again for the lovely collaboration!

gimenete reviewed (commented):
Thanks for addressing the feedback! I left some additional comments.

One nitpick: some bullet points start with upper case while others start with lower case.

TejasQ commented:
Alrighty, all comments addressed. 😬 Thanks for a great review. Looking forward to another round when convenient.

gimenete reviewed (approved):
Looks great! ✨ 

Thanks for addressing all the feedback! 🙏
